### DISCUSSION TITLE
How to render page as a string

### DISCUSSION THREAD

**Discussion by martin-dimi** [vercel/next.js#41122] [Help]
Upon calling a nextjs API, I need to render a page onto a string which can be manipulated. For instance, something like this:

```
export default function handler(req, res) {
    const props = req.body.props()
    const html = next.renderDocument(<SomePage ...props>);  // HOW? or 
    doSomething(html);
    res.status(200);
}
```

I tried using ReactDOMServer.renderToString(<SomePage>) but that doesn't work as it only renders the components but not the whole page (e.g. doesn't add stylesheet, HTML tag etc). I was hoping there would be a document call I can make but am struggling to find one.

Is this even possible? How can I achieve this?

**Comment by devjiwonchoi** [comment]
You can use `dangerouslySetInnerHTML`.

### API

```ts
// /api/render.ts
import { NextApiRequest, NextApiResponse } from 'next'

export default function handler(req: NextApiRequest, res: NextApiResponse) {
  const html = '<h1>Hello World!</h1>'

  res.status(200).json({ html })
}
```

### Page

```ts
// /index.tsx

export default function Home({ html }: { html: string }) {
  return (
    <main>
      <div dangerouslySetInnerHTML={{ __html: html }}></div>
    </main>
  )
}

export async function getServerSideProps() {
  const res = await fetch(`http://localhost:3000/api/render`)
  const data = await res.json()
  const { html } = data
  return { props: { html } }
}
```

![Screenshot 2023-09-06 at 8 36 04 AM](https://github.com/vercel/next.js/assets/120007119/84d3a93e-ec19-4047-87c8-996f0f394ff8)

**Comment by Maclay74** [comment]
@martin-dimi Hey, have you resolved it somehow?

**Reply by garciadiazjaime** [reply]
hey @Maclay74 I shifted to Express + React due to the project's requirement to support very old browser versions. Next.js didn't work, and trying to polyfill for compatibility became a time-consuming rabbit hole.

I'm doing something like this:
```
app.get("/url", async (req, res) => {

const pageHTML = ReactDOM.renderToString(<Page data={data} />);

  res.send(
    `<html>
      <body>
        <div id="root">${pageHTML}</div>
        <script src="/static/bundle.js"></script>
      </body>
    </html>`
  );
```

hope it helps

**Reply by Maclay74** [reply]
Yep, thanks a lot!

**Reply by martin-dimi** [reply]
Not through nextjs I'm afraid. Have a similar setup to what @garciadiazjaime suggested

**Comment by devodii** [comment]
Hey, so no fix for this yet? :/
